File: Automation_project/driver/imagecomp.pl
#!/usr/local/bin/perl
use Win32::Registry;
use Win32::OLE;
$classpath="\\\\bigrepos01\\bigscm\\admin\\bin";
use Cwd;

$workspace=getcwd();
$workspace=~ s/\//\\/;
$br = shift(@ARGV);
$suitepath = shift(@ARGV);
#$testrailpath = shift(@ARGV);
$list= shift(@ARGV);
$suite = $suitepath ;
$suite=~s /.*\\//;
$imageDir="$suitepath\\images";
exit
  if not (-e $imageDir);
$actual_imageDir="$suitepath\\actual_images";
chdir($imageDir);
open ( FILE,">C:\\temp\\${suite}.html");

print FILE "<html>\n";
print FILE "<head>\n";
print FILE "<title>image comparison report</title>\n";
print FILE "</head>\n";
print FILE  "<body>\n";
$time=localtime(time);
print FILE "<h1>Report for $suite in $br-- Date:${time}</h1>\n";
print FILE "<script type=\"text/javascript\">\n";
print FILE  "function compare(pic1,pic2,reportname){\n";
print FILE "var w = new ActiveXObject(\"WScript.Shell\")\n";
print FILE "w.run (\"\\\\\\\\bigrepos01\\\\bigscm\\\\tst_results\\\\compare2pics.vbs \"+pic1+\" \" +pic2+\" \"+reportname)\n";
print FILE "}\n";
print FILE "</script>\n";
print FILE "<table>\n";


my @resp=qx(ls -a|sort);
   
   
foreach(@resp) {
  
 if($_ =~ /C\d+/){
  
  chomp($_);
  $base=$_;
  $actual=$_;
  $testname=$base;
  $testname =~ s/\_.*//;
  chomp ($testname);
  $actual =~ s/Base/Actual/;
  $baseimage="$imageDir\\${base}";
  $newimage="$actual_imageDir\\${actual}";
  #print "$baseimage\n";
  if( -e "$workspace/$newimage"){
  #print "$newimage\n";
print FILE "<tr>\n";
$cmd="node c:\\compare2pics.js $workspace/${baseimage} $workspace/$newimage $testname";
$status= qx($cmd);
my @now = localtime();
my $timeStamp = sprintf("%04d%02d%02d%02d%02d%02d", 
                        $now[5]+1900, $now[4]+1, $now[3],
                        $now[2],      $now[1],   $now[0]);

$baseimage_save=$baseimage;
$newimage_save=$newimage;
$baseimage_save=~ s/\.png//i;
$baseimage_save=~ s/.*\\//i;
$newimage_save=~ s/\.png//i;
$newimage_save=~ s/.*\\//i;

$baseimage_save="\\\\bigrepos01\\bigscm\\tst_results\\${baseimage_save}_${timeStamp}".".png";
$newimage_save= "\\\\bigrepos01\\bigscm\\tst_results\\${newimage_save}_${timeStamp}".".png";
$cmd="copy /y  $workspace\\${baseimage} $baseimage_save";
print "$cmd\n";
system($cmd);
$cmd="copy /y  $workspace\\$newimage $newimage_save";
print "$cmd\n";
system($cmd);
$baseimage_save=~s/\\/\\\\/g;
$newimage_save=~s/\\/\\\\/g;



print FILE"<td><A href=\"javascript:compare('$baseimage_save','$newimage_save','$testname')\">$testname</A></td>\n";
if  ($status eq "pass"){ 
    print FILE "<td align=\"right\"><font color=green>Passed</td>\n";
}elsif ($status = "fail"){
    print FILE "<td align=\"right\"><font color=red>Failed</td>\n";
}
}

}
print FILE "</tr>\n";
}

print FILE "</br\n";
print FILE "</table>\n";
print FILE "</body>\n";
print FILE "</html>\n";
close(FILE);
$cmd="phantomjs c:\\testrail.js $suite \"$testrailpath\"";
print "$cmd\n";
#system($cmd);
$mailgp="Arif_Bhuiyan\@ibi.com,$list";
$contacts="$mailgp".","."$contacts";
$from="qaauto\@ibi.com";
$sub="\"$suite image comparison report from $br\"";
$message="\"Please save ${suite}.html before open in ie\"";
$report1="c:\\temp\\${suite}.html";
#$report2="c:\\temp\\${suite}.png";
$cmd="java   -cp $classpath emailtogp $contacts $from  $sub $message $report1 $report2";
print "$cmd\n";
system($cmd);



# call ie 

#$ie = Win32::OLE->new('InternetExplorer.Application') ;
#$ie->{'Visible'} = 1;
#$ie->navigate("C:\\resutlt.html");
#$ie->Refresh();

File: Automation_project/driver/imagecomp.py
import glob, os, re, sys, subprocess
from PIL import Image
from PIL import ImageChops

def compare_images(test, path_one, path_two, diff_save_location, log_file):
    """
    Compares to images and saves a diff image, if there
    is a difference
 
    @param: path_one: The path to the first image
    @param: path_two: The path to the second image
    """
    fileObj = open(log_file, "a")
    try:
        image_one = Image.open(path_one)
    except FileNotFoundError:
        fileObj.write(test + " file is not available in Base location.")
        fileObj.write("\n")
        return
    image_two = Image.open(path_two)
    one=image_one.convert('RGBA')
    two=image_two.convert('RGBA')
    
    diff = ImageChops.difference(one, two)
    
    if diff.getbbox():
        diff.save(diff_save_location)
        fileObj.write(test + " - Failed.")
        fileObj.write("\n")
    else:
        fileObj.write(test + " - Passed.")
        fileObj.write("\n")
    fileObj.close()
    
def send_email(suite, mail_list, diff_report_location):
    classpath="\\\\bigrepos01\\bigscm\\admin\\bin";
    mailto=mail_list
    mailfrom="\"qaauto@ibi.com\""
    sub="\"Image comparison Report for " + suite + ".\""
    message="\"Please Open the attachment to find the image comparison Report.\""
    diff_report=diff_report_location
    cmd="java  -cp " + classpath + ' emailtogp \"' + mailto + '\" ' + mailfrom + " " + sub + " " + message + ' \"' + diff_report + '\"'
    subprocess.Popen(cmd)
    
"""
Here is the main program.
Syntax: python imagecomp.py "D:\wf3_S9164_pdf_chart_1_cr\qa\selenium\S9164_pdf_chart_1" "se.automation@amtexsystems.com"
"""
suite=sys.argv[1]
mail_list=sys.argv[2]
root=os.getcwd()
basepath=root + "\\qa\\selenium\\" + suite
actual_image_path=basepath + "\\actual_images\\"
base_image_path=basepath + "\\images\\"
diff_image_path=basepath + "\\diff\\"
log_file=basepath + "\\diff\\image_compare.txt"

if not os.path.exists(diff_image_path):
    os.makedirs(diff_image_path)
if len(glob.glob(actual_image_path + "\\*.*")) > 1:
    for name in glob.glob(actual_image_path + "\\*.*"):
        if bool('actual.png' not in name):
            try:
                test=re.match('.*([C|c]\d+.*\.png)', name).group(1)
            except AttributeError:
                print(name+" file not found.")
            base=re.sub('Actual', 'Base', test)
            diff=re.sub('Actual', 'Diff', test)
            diff=re.sub('png', 'jpg', diff)
            base=base_image_path + base
            act=actual_image_path + test
            diff=diff_image_path + diff
            compare_images(test, base, act, diff, log_file)
    send_email(suite, mail_list, log_file)
else:
    print("There is no file to compare.")
